flink-runtime/src/main/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGenerator.java

```diff
@@ -355,32 +337,44 @@ private void waitForSerializationFuturesAndUpdateJobVertices()
         }
     }
 
-    private void addVertexIndexPrefixInVertexName() {
-        if (!streamGraph.isVertexNameIncludeIndexPrefix()) {
+    // Set this property for the newly generated JobVertex
+    public static void addVertexIndexPrefixInVertexName(
+            JobVertexBuildContext jobVertexBuildContext,
+            AtomicInteger vertexIndexId,
+            JobGraph jobGraph) {
+        if (!jobVertexBuildContext.getStreamGraph().isVertexNameIncludeIndexPrefix()) {
             return;
         }
-        final AtomicInteger vertexIndexId = new AtomicInteger(0);
+        Set<JobVertexID> jobVertexIds =
+                jobVertexBuildContext.getJobVertices().values().stream()
+                        .map(JobVertex::getID)
+                        .collect(Collectors.toSet());
         jobGraph.getVerticesSortedTopologicallyFromSources()
                 .forEach(
-                        vertex ->
+                        vertex -> {
+                            if (jobVertexIds.contains(vertex.getID())) {
                                 vertex.setName(
                                         String.format(
                                                 "[vertex-%d]%s",
-                                                vertexIndexId.getAndIncrement(),
-                                                vertex.getName())));
+                                                vertexIndexId.getAndIncrement(), vertex.getName()));
+                            }
+                        });
     }
```

Comment:
It seems that this method does not need to consider the concept of newly added vertices; you can simply retrieve the vertices from the context and set them directly. However, in the case of incremental graph, the context only contains the incremental vertices, while the setName method needs to set based on the entire topology. Therefore, it has to rely on the job graph to calculate a difference set.

I suggest updating the Javadoc for this method.